### PR TITLE
Support updating federated role

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

Fixes https://issues.redhat.com/browse/OSD-4235

Usage

```bash
osdctl federatedrole apply -u <URL>

osdctl federatedrole apply -f <yaml file>
```

### CODE REVIEW

File: cmd/federatedrole/apply.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+			federatedAccount.Spec.AWSFederatedRole.Name == federatedRole.Name {
+			if accountID, ok = federatedAccount.Labels[awsAccountIDLabel]; !ok {
+				return errors.New("Unable to get AWS AccountID label for AWS federated account access CR " + federatedAccount.Name)
+			}
+
+			if uid, ok = federatedAccount.Labels[uidLabel]; !ok {
+				return errors.New("Unable to get UID label for AWS federated account access CR " + federatedAccount.Name)
+			}
+
+			roleArn := fmt.Sprintf("arn:aws:iam::%s:role/%s", accountID, awsv1alpha1.AccountOperatorIAMRole)
```
Comment (fahlmant) [inline cmd/federatedrole/apply.go:173]: The role here should have the uid appended

**Comment by yeya24** [comment]
@fahlmant One thing I am not sure, do I need to update the specified federated role cr in this command? 

If not, users need to do something like `oc apply -f <file>` and then `osdctl federatedrole apply -f <file>`

File: pkg/provider/aws/iam.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	Principal *awsv1alpha1.Principal `json:"Principal,omitempty"`
+}
+
+type policyDoc struct {
+	Version   string         `json:"Version"`
+	Statement []awsStatement `json:"Statement"`
+}
+
+func RefreshIAMPolicy(awsClient Client, federatedRole *awsv1alpha1.AWSFederatedRole, awsAccountID, uid string) error {
+	roleName := federatedRole.Name + "-" + uid
```
Comment (fahlmant) [inline pkg/provider/aws/iam.go:26]: Ah I see you append the uid to the role here instead

File: pkg/provider/aws/iam.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() != iam.ErrCodeNoSuchEntityException {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	if _, err := awsClient.CreatePolicy(&iam.CreatePolicyInput{
```
Comment (fahlmant) [inline pkg/provider/aws/iam.go:71]: we may need some sort of retry with exponential back off in the case that the API gives the EntityAlreadyExists err.

**Comment by fahlmant** [comment]
@yeya24 If you look at the done criteria, that is outlined in the first point
```
osdctl can grab a Federated Role CR at a supplied url and apply it
```

File: cmd/federatedrole/apply.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+				AwsAccessKey: *credentials.SecretAccessKey,
+				AwsToken:     *credentials.SessionToken,
+				AwsRegion:    o.region,
+			})
+			if err != nil {
+				return err
+			}
+
+			fmt.Fprintln(o.Out, "Updating IAM policy for AWS federated account access CR "+federatedAccount.Name)
+			if err := awsprovider.RefreshIAMPolicy(awsClient, &federatedRole, accountID, uid); err != nil {
```
Comment (fahlmant) [inline cmd/federatedrole/apply.go:191]: One case I just thought of is the FederatedRole CR may not have a new custom policy but have a new managed Policy in the list. So the code should also ensure that all the Managed policies specified are attached to the Role and detached any that may have been removed

File: pkg/provider/aws/iam.go (hunk 3 above)
Comment (yeya24) [reply pkg/provider/aws/iam.go:71]: Do you mean backoff retry? But I see we are doing a different thing here https://github.com/openshift/aws-account-operator/blob/master/pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go#L431

**Comment by tfahlman** [comment]
> I'm not sure how best to test this. @tfahlman, do you mind taking a look here again when you get a chance?

I think you tagged the wrong person :)

**Comment by clcollins** [comment]
Doh!  I meant to tag @fahlmant!

Same question: I'm not sure how to test this - do you mind taking a look here again when you get a chance?

File: cmd/account/console.go [hunk 5]
```diff
@@ -120,7 +120,7 @@ func (o *consoleOptions) run() error {
 		return err
 	}
 	if o.verbose {
-		fmt.Println(callerIdentityOutput)
+		fmt.Fprintln(o.Out, callerIdentityOutput)
```
Comment (fahlmant) [inline cmd/account/console.go:123]: What's this doing?

File: cmd/account/get/cmd.go [hunk 6]
```diff
@@ -14,7 +14,7 @@ const (
 func NewCmdGet(streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) *cobra.Command {
 	getCmd := &cobra.Command{
 		Use:               "get",
-		Short:             "get resources",
+		Short:             "Get resources",
```
Comment (fahlmant) [inline cmd/account/get/cmd.go:17]: Does this need to be included?

File: docs/command/osdctl_account.md [hunk 7]
```diff
@@ -34,7 +34,7 @@ osdctl account [flags]
 * [osdctl account clean-velero-snapshots](osdctl_account_clean-velero-snapshots.md)	 - Cleans up S3 buckets whose name start with managed-velero
 * [osdctl account cli](osdctl_account_cli.md)	 - Generate temporary AWS CLI credentials on demand
 * [osdctl account console](osdctl_account_console.md)	 - Generate an AWS console URL on the fly
-* [osdctl account get](osdctl_account_get.md)	 - get resources
+* [osdctl account get](osdctl_account_get.md)	 - Get resources
```
Comment (fahlmant) [inline docs/command/osdctl_account.md:37]: Is this related?

Comment (yeya24) [reply docs/command/osdctl_account.md:37]: Umm, not related. I can remove all the unrelated changes if you think that's better. @fahlmant
